### PR TITLE
Add workspace maximum time

### PR DESCRIPTION
**Pull request by tomgeorge**
### What does this PR do?

Adds the property `che.workspace.limits.run.timeout` to set the maximum amount of time a workspace can run before it is stopped by the activity manager, regardless of activity

### What issues does this PR fix or reference?

#17231

### CODE REVIEW

**Comment by tomgeorge** [comment]
WIP because a test is failing, but I wanted to get opinions on the approach, naming, etc.

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties
Comment (sleshchenko) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: 1. If I'm not mistaken `che.limits.workspace.idle.timeout` is just a default value applied to every user, but then admin is able to override it per user. Does it make sense to implement the same for run.timeout?
2. From UX perspective I suppose it would be nice to see a warning when like there are 3 minutes left before workspace is shutdown...

File: wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivity.java
Comment (skabashnyuk) [inline wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivity.java]: why do we need extra field if we have already  `   last_starting              BIGINT and   last_running               BIGINT`, which can be used for how long workspace is already running?

File: wsmaster/che-core-sql-schema/src/main/resources/che-schema/7.16.0/2__update_che_workspace_activity_add_hard_time_limit.sql
Comment (skabashnyuk) [inline wsmaster/che-core-sql-schema/src/main/resources/che-schema/7.16.0/2__update_che_workspace_activity_add_hard_time_limit.sql]: if we decided to keep this field then we need to think about extending `CREATE INDEX che_index_ws_activity_expiration ON che_workspace_activity (expiration);` to make `WorkspaceActivity.getExpired` query efficient.

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties
Comment (skabashnyuk) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: ++

Comment (ibuziuk) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: @tomgeorge `che.limits.workspace.run.timeout`  Should be disabled by default 

@sleshchenko 

1. I was not aware of the fact that `che.limits.workspace.idle.timeout` can be set per-user. Can you point me to the docs if exist? In general, per-user settings for `che.limits.workspace.run.timeout` can be adressed in a separate issue if needed.

2. it is planned to be disabled by default and on Hosted Che set smth. like 5 / 8 h mainly to prevent abusers so  I do not think UX would be affected by this. In general, UX should be improved for vanilla idling in the first place

Comment (sleshchenko) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: @ibuziuk 

> 1. I was not aware of the fact

Fair point! Even that property comment does not mention that it's possible to override it per user with Resources API.
So, unfortunately that documentation is not ported from Che 6 but luckily I found it https://github.com/eclipse/che-docs/blob/2ca44011114cce9ef93963d633950289548371ce/src/main/pages/che-6/user-management/resource-management.adoc
The limits users have are just default values which can be overridden by Che Admin via REST API
Screenshot_20200703_164313

> In general, per-user settings for che.limits.workspace.run.timeout can be adressed in a separate issue if needed.

I don't have any issues with that :+1: 

> 2. it is planned to be disabled by default and on Hosted Che set smth. like 5 / 8 h mainly to prevent abusers so I do not think UX would be affected by this. In general, UX should be improved for vanilla idling in the first place

Makes sense, I didn't dig deep enough to get the idea why we had that limit. If it's really big like 5-8h - then I agree that UX would be affected too much.

Comment (tomgeorge) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: > From UX perspective I suppose it would be nice to see a warning when like there are 3 minutes left before workspace is shutdown...

Is it correct that this would require a che-theia change?

Comment (sleshchenko) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: > Is it correct that this would require a che-theia change?

I suppose - yes. But as discussed with Illya - the only thing that makes sense to do with that in the current PR - create an issue for CheServer/Dashboard/Che Theia to improve UX when workspace is stopped by run_timeout or idle_timeout.

**Comment by tomgeorge** [comment]
Removed database column and now compares current time - last running > run timeout

**Comment by gazarenkov** [comment]
Do we need to update documentation ? 
Provide a link to Docs PR if so

File: wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivity.java
Comment (skabashnyuk) [inline wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivity.java]: 1. it would be interesting to see `EXPLAIN` of this query to figure out missing indexes.
2. :runTimeout > 0 - I think it would be better to split this query on two.  Orginal `SELECT a FROM WorkspaceActivity a WHERE a.expiration < :expiration` and current and execute it depending of the value of `runTimeout`
3. if runTimeout > 0 query may look like `WHERE a.expiration < :expiration OR (a.status = org.eclipse.che.api.core.model.workspace.WorkspaceStatus.RUNNING AND a.lastRunning< :X)  I assume it would be easier for DB to optimize. Anyway we need to see `EXPLAIN`

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties
Comment (ibuziuk) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: IMO it worth clarifying a couple of things:
- 0 = disabled by default
- what time unit is expected to be used in the property (milliseconds probably ? similar to what is used in the `che.limits.workspace.idle.timeout` )

Comment (ibuziuk) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: the proper detailed explanation of the property is important since it will automatically generate the docs - https://www.eclipse.org/che/docs/che-7/advanced-configuration-options-for-the-che-server-component/#che-server-component-system-properties-reference_advanced-configuration-options-for-the-che-server-component

Comment (tomgeorge) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/multiuser.properties]: Updated the documentation to clarify these points

**Comment by tomgeorge** [comment]
@gazarenkov it looks like the documentation for this property will be automatically generated

**Comment by tomgeorge** [comment]
crw-ci-test

**Comment by ibuziuk** [comment]
> Do we need to update documentation ? Provide a link to Docs PR if so

@gazarenkov I would expect that docs will be automatically generated based on the property file  - https://www.eclipse.org/che/docs/che-7/advanced-configuration-options-for-the-che-server-component/#che-server-component-system-properties-reference_advanced-configuration-options-for-the-che-server-component

**Comment by tomgeorge** [comment]
[crw-ci-test]

**Comment by tomgeorge** [comment]
[crw-ci-test]

**Review by ibuziuk — APPROVED** [review]
LGTM, please follow up with @skabashnyuk about the query implementation

File: wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivity.java
Comment (tomgeorge) [reply wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivity.java]: Expain results

old
```
                                                    QUERY PLAN                                                    
------------------------------------------------------------------------------------------------------------------
 Index Scan using che_index_ws_activity_expiration on che_workspace_activity a  (cost=0.41..8.10 rows=1 width=82)
   Index Cond: (expiration < '1594228056124'::bigint)
(2 rows)


real	0m0.077s
user	0m0.040s
sys	0m0.020s

```

new
```
                                                                     QUERY PLAN                                                                      
-----------------------------------------------------------------------------------------------------------------------------------------------------
 Bitmap Heap Scan on che_workspace_activity a  (cost=8.88..33.34 rows=2 width=82)
   Recheck Cond: ((expiration < '1594228056124'::bigint) OR ((status)::text = 'RUNNING'::text))
   Filter: ((expiration < '1594228056124'::bigint) OR (((status)::text = 'RUNNING'::text) AND (('1594228056124'::bigint - last_running) > 3600000)))
   ->  BitmapOr  (cost=8.88..8.88 rows=7 width=0)
         ->  Bitmap Index Scan on che_index_ws_activity_expiration  (cost=0.00..4.42 rows=1 width=0)
               Index Cond: (expiration < '1594228056124'::bigint)
         ->  Bitmap Index Scan on che_index_ws_activity_last_stopped  (cost=0.00..4.46 rows=7 width=0)
               Index Cond: ((status)::text = 'RUNNING'::text)
(8 rows)


real	0m0.080s
user	0m0.040s
sys	0m0.018s

```

Comment (ibuziuk) [reply wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivity.java]: @tomgeorge actually I forgot one important thingy - ideally, we want to have a log message when a workspace is stopped by the `che.limits.workspace.run.timeout` in order to identify and investigate this case which is likely to be an abuser. so @skabashnyuk proposal about splitting query into two is relevant

File: wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivity.java
Comment (skabashnyuk) [inline wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivity.java]: - remove `:runTimeout > 0` condition
- replace `:timestamp - :runTimeout > a.lastRunnin` with `a.lastRunnin<X`

File: wsmaster/che-core-api-workspace-activity/src/test/java/org/eclipse/che/api/workspace/activity/spi/tck/WorkspaceActivityDaoTest.java
Comment (skabashnyuk) [inline wsmaster/che-core-api-workspace-activity/src/test/java/org/eclipse/che/api/workspace/activity/spi/tck/WorkspaceActivityDaoTest.java]: best practice in this part of the code is to avoid wildcard imports.

File: wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivityChecker.java
Comment (ibuziuk) [inline wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivityChecker.java:129]: @tomgeorge the reason will be logged in the stopWorkspace e.g. (Workspace run timeout exceeded), right?

Comment (ibuziuk) [inline wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivityChecker.java:125]: @tomgeorge the reason will be logged in the stopWorkspace e.g. ("Workspace run timeout exceeded"), right?

**Review by amisevsk — APPROVED** [review]
LGTM but I defer to @skabashnyuk's comments on the queries.

It's a bit strange to force workspace idling without warning, to me, at least for a regular user who may be actively using their workspace, but that is avoided by setting a sensible timeout I suppose.

**Comment by tomgeorge** [comment]
> It's a bit strange to force workspace idling without warning, to me, at least for a regular user who may be actively using their workspace, but that is avoided by setting a sensible timeout I suppose.

Yes, I agree.  The default timeout is zero, so unless you explicitly set it, it will be off.  I think on the UI side we'll implement a warning notification that the workspace will soon be stopped.

File: wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivity.java
Comment (skabashnyuk) [inline wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivity.java]: - select  a.workspaceId? I think we can make same replacement in  WorkspaceActivity.getExpiredIdle too"

-  replace two parameters query  `:timestamp - :runTimeout` with one parameter `:X` do not ask the database to make additional calculations that are not needed.
Can you provide `explain` of this query, please? Do we need an index on `a.status, a.lastRunning`?

Comment (tomgeorge) [reply wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivity.java]: I will change to return a list of strings instead of `WorkspaceActivity`, and will pass in the difference to the query.

See the above explain plan.  The existing indexes should be fine.

File: wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivityChecker.java
Comment (skabashnyuk) [inline wsmaster/che-core-api-workspace-activity/src/main/java/org/eclipse/che/api/workspace/activity/WorkspaceActivityChecker.java:116]: what if we add a parameter `reason` to `stopExpiredQuietly` then this lambda will look a bit smaller.?

**Review by skabashnyuk — APPROVED** [review]
you good to merge if both queries correctly use db indexes.

**Comment by skabashnyuk** [comment]
Don't forget to add/check documentation.

**Comment by tomgeorge** [comment]
[ci-build]

**Comment by ibuziuk** [comment]
[ci-build]

**Pull request merged**
